src/nix/narinfo.rs

```diff
@@ -0,0 +1,246 @@
+    ///
+    /// # Arguments
+    ///
+    /// * `narinfo` - A string slice that holds the `narinfo` data.
+    ///
+    /// # Returns
+    ///
+    /// * `Result<NarInfo, Error>` - Returns a `NarInfo` struct if parsing is successful, otherwise returns an `Error`.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if:
+    /// * A line in the `narinfo` string does not contain a ':' delimiter.
+    /// * The `narinfo` string contains an unknown key.
+    /// * A required field is missing from the `narinfo` string.
+    pub fn parse(narinfo: &str) -> crate::error::Result<Self> {
+        let mut store_path = None;
+        let mut url = None;
+        let mut compression = None;
+        let mut file_hash = None;
+        let mut file_size = None;
+        let mut nar_hash = None;
+        let mut nar_size = None;
+        let mut deriver = None;
+        let mut system = None;
+        let mut references = Vec::new();
+        let mut sig = None;
+        let mut ca = None;
+
+        for line in narinfo.lines() {
+            let (key, value) = line
+                .split_once(':')
+                .ok_or_else(|| crate::error::Error::NarInfoNoDelimiter(line.to_string()))
+                .map(|(key, value)| (key.trim(), value.trim()))?;
+
+            match key {
+                "StorePath" => store_path = Some(value.to_string()),
+                "URL" => url = Some(value.to_string()),
+                "Compression" => compression = Some(value.to_string()),
+                "FileHash" => file_hash = Some(value.to_string()),
+                "FileSize" => file_size = Some(value.parse()?),
+                "NarHash" => nar_hash = Some(value.to_string()),
+                "NarSize" => nar_size = Some(value.parse()?),
+                "Deriver" => deriver = Some(value.to_string()),
+                "System" => system = Some(value.to_string()),
+                "References" => references = value.split(' ').map(|s| s.to_string()).collect(),
+                "Sig" => sig = Some(value.to_string()),
+                "CA" => ca = Some(value.to_string()),
+                _ => return Err(crate::error::Error::NarInfoUnknownKey(key.to_string())),
+            }
+        }
+
+        Ok(NarInfo {
+            store_path: store_path
+                .ok_or_else(|| crate::error::Error::NarInfoMissingField("StorePath".to_string()))?,
+            url: url.ok_or_else(|| crate::error::Error::NarInfoMissingField("URL".to_string()))?,
+            compression: compression.ok_or_else(|| {
+                crate::error::Error::NarInfoMissingField("Compression".to_string())
+            })?,
+            file_hash: file_hash
+                .ok_or_else(|| crate::error::Error::NarInfoMissingField("FileHash".to_string()))?,
+            file_size: file_size
+                .ok_or_else(|| crate::error::Error::NarInfoMissingField("FileSize".to_string()))?,
```

Comment:
These also don't seem to be mandatory in Nix

Reply:
Is there a place where I can concretely see which fields are and aren't expected by Nix? I could ofc just assume all are optional, but that limited what kind of stuff we can do with them.

Reply:
My main source was https://github.com/nixos/nix/blob/0d26358bda1637e54ebacac18e7b5af7381cf5f3/src/libstore/nar-info.cc#L98-L106

Reply:
Thank you <3